### PR TITLE
gh-99633: Add context manager support to `contextvars.Context`

### PR DESCRIPTION
**Pull request by rhansen**
* Issue: gh-99633

### CODE REVIEW

File: Python/context.c [hunk 1]
```diff
@@ -558,6 +557,47 @@ context_tp_contains(PyContext *self, PyObject *key)
 }
 
 
+/*[clinic input]
+_contextvars.Context.__enter__
+
+Context manager enter.
```
Comment (sobolevn) [inline Python/context.c:787]: This will go into `__doc__` of these methods. Probably it can be something more useful. Like: a short sentence about when it should be used, maybe even mentioning `.run` alternative.

File: Python/context.c [hunk 2]
```diff
@@ -558,6 +557,47 @@ context_tp_contains(PyContext *self, PyObject *key)
...
+static PyObject *
+_contextvars_Context___exit___impl(PyContext *self, PyObject *exc_type,
+                                   PyObject *exc_val, PyObject *exc_tb)
+/*[clinic end generated code: output=4608fa9151f968f1 input=31b822b221c4439b]*/
+{
+    PyThreadState *ts = _PyThreadState_GET();
+    if (_PyContext_Exit(ts, (PyObject *)self)) {
+        return NULL;
+    }
+    Py_RETURN_FALSE;
```
Comment (sobolevn) [inline Python/context.c]:
```suggestion
    Py_RETURN_NONE;
```

File: Python/context.c [hunk 3]
```diff
@@ -558,6 +557,47 @@ context_tp_contains(PyContext *self, PyObject *key)
...
+
+static PyObject *
+_contextvars_Context___enter___impl(PyContext *self)
+/*[clinic end generated code: output=7374aea8983b777a input=093b5c4808ddf1b1]*/
+{
+    PyThreadState *ts = _PyThreadState_GET();
+    if (_PyContext_Enter(ts, (PyObject *)self)) {
+        return NULL;
+    }
+    return Py_NewRef(self);
```
Comment (sobolevn) [inline Python/context.c:832]: Reviewing note: `self` object is incremented twice. Here and in `_PyContext_Enter`: `ts->context = Py_NewRef(ctx);`

File: Lib/test/test_context.py [hunk 4]
```diff
@@ -360,6 +361,62 @@ def sub(num):
...
+        with self.assertRaises(RuntimeError):
+            ctx.__exit__(None, None, None)
+
+    def test_context_manager_3(self):
+        with contextvars.copy_context() as ctx:
+            with self.assertRaises(RuntimeError):
+                with ctx:
+                    pass
+
+    def test_context_manager_4(self):
```
Comment (sobolevn) [inline Lib/test/test_context.py]: Let's add more semantic names to these tests.
For example, this one can be: `test_run_inside_context_raises`.

However, this is the naming style that is used in `test_context` right now.

File: Lib/test/test_context.py [hunk 5]
```diff
@@ -360,6 +361,62 @@ def sub(num):
             tp.shutdown()
         self.assertEqual(results, list(range(10)))
 
+    @isolated_context
+    def test_context_manager_1(self):
+        cvar = contextvars.ContextVar('cvar', default='initial')
+        self.assertEqual(cvar.get(), 'initial')
+        ctx = contextvars.copy_context()
+        with ctx as tmp:
+            self.assertIs(tmp, ctx)
```
Comment (sobolevn) [inline Lib/test/test_context.py]: Do we really need to return `tmp` from `__enter__` here? We now have two identical identities with different names. It might confuse users.

File: Lib/test/test_context.py [hunk 6]
```diff
@@ -360,6 +361,62 @@ def sub(num):
...
+            self.assertEqual(cvar.get(), 'updated')
+        self.assertEqual(cvar.get(), 'initial')
+        self.assertEqual(ctx[cvar], 'updated')
+        with ctx:
+            self.assertEqual(cvar.get(), 'updated')
+
+    @threading_helper.requires_working_threading()
+    def test_context_manager_2(self):
+        ctx = contextvars.copy_context()
+        thread = threading.Thread(target=ctx.__enter__)
```
Comment (sobolevn) [inline Lib/test/test_context.py:419]: I think it won't hurt to add a thread-safety test, something like `test_context_threads_1`. How hard would it be?

**Review by sobolevn — COMMENTED** [review]
Thank you! Looks interesting :)

File: Lib/test/test_context.py (hunk 6 above)
Comment (rhansen) [reply Lib/test/test_context.py:419]: I don't quite understand what `test_context_threads_1` is meant to test.  Maybe it's testing that each thread automatically gets its own Context instance?  If threads didn't automatically get their own context then that test would fail with high probability, I think.

If that test is meant to test internal state data races then I don't understand how it accomplishes that goal, as sleeping for even 1ms significantly drops the probability of losing a data race.

It would be great to write a test that guarantees that concurrent `Context.__enter__()` and/or `Context.run()` calls on the same Context object will result in exactly one successful call, and `RuntimeError` exceptions for the other calls, without any internal state corruption due to data races.  I'm not sure how to write such a test.

However, I'm not sure it's worth adding such a test in this PR:
  * The possibility of concurrent enters is not new to this PR (one can call `Context.run()` concurrently today).
  * This PR doesn't make concurrent enters more likely.
  * Users are unlikely to encounter a race because Context objects are not meant to be used across threads.  (There is no way to hand off ownership of an entered Context to another thread, and exiting a Context from a different thread will raise.)
  * I think the test would fail because `_PyContext_Enter` does not atomically check and set `ctx->ctx_entered`.  Fixing that feels out of scope for this PR.

It might be worth addressing thread safety in another PR, but as far as I know nobody has complained so far, so maybe it's better to just leave it alone.

File: Python/context.c (hunk 3 above)
Comment (rhansen) [reply Python/context.c:832]: Yes.  The refcount added in `_PyContext_Enter` is subtracted in `_PyContext_Exit` by the (oddly named) `Py_SETREF` macro.

Both increments are necessary:  one is to keep the Context alive while it is entered, and one is for the `with` statement's `as` binding.

File: Python/context.c (hunk 1 above)
Comment (rhansen) [reply Python/context.c:787]: Done.

File: Python/context.c (hunk 2 above)
Comment (rhansen) [reply Python/context.c]: Done.

File: Lib/test/test_context.py (hunk 4 above)
Comment (rhansen) [reply Lib/test/test_context.py]: Done.  I also split the first test into three to clarify intentions.

File: Lib/test/test_context.py (hunk 5 above)
Comment (rhansen) [reply Lib/test/test_context.py]: Addressed by splitting the `assertIs` check to its own test case.

**Comment by cjw296** [comment]
@sobolevn - do you still have outstanding concerns?

I think we'll have to assume @1st1 is not going to get to this PR and it would be great to have it landed.

**Comment by picnixz** [comment]
Before merging it, I'd like to have a look (probably tomorrow).

**Comment by sobolevn** [comment]
@cjw296 no, I don't have any concerns :)
Please, take this over!

**Comment by cjw296** [comment]
@picnixz - great! I'd love more eyes on making sure this really has the same effect as `copy_context()` and then having the body on the managed context behave equivalently to being within a `ctx.run()`!

File: Doc/library/contextvars.rst [hunk 7]
```diff
@@ -144,21 +144,89 @@ Manual Context Management
...
+   method are recorded in the current Context.  The :meth:`ContextVar.get`
+   method returns the value associated with the current Context.  Thus, exiting
+   a Context effectively reverts any changes made to context variables while the
+   Context was entered.  (If desired, the values can be restored by re-entering
+   the Context.)
 
    Context implements the :class:`collections.abc.Mapping` interface.
 
+   .. versionadded:: 3.14
+      Context implements the :term:`!context management protocol` (:pep:`343`).
```
Comment (picnixz) [inline Doc/library/contextvars.rst]:
```suggestion
      Added support for the :term:`context management protocol`.
```

Let's reduce the number of links.

File: Doc/library/contextvars.rst [hunk 8]
```diff
@@ -144,21 +144,89 @@ Manual Context Management
...
+   .. method:: __enter__(self)
+               __exit__(self, exc_type, exc_value, exc_tb)
+
+      These methods implement the :term:`!context management protocol`
+      (:pep:`343`), making it possible to use a Context object as a
+      :term:`context manager` with the :keyword:`with` statement.  The
+      :meth:`__enter__` method enters the Context and returns *self*, so the
+      value bound to the identifier given in the :keyword:`with` statement's
+      :keyword:`!as` clause (if present) is the Context object itself.  The
+      :meth:`__exit__` method exits the Context.
```
Comment (picnixz) [inline Doc/library/contextvars.rst]:
```suggestion
```

Other classes don't seem to document those kind of methods. Since they don't do much more than what you already explained, we can simplify the docs (but keep the examples).

File: Doc/library/contextvars.rst [hunk 9]
```diff
@@ -144,21 +144,89 @@ Manual Context Management
...
+         # Exited ctx, so the observed value of var has reverted.
+         assert var.get() == "initial"
+         # But the updated value is still recorded in ctx.
+         assert ctx[var] == "updated"
+
+         # Re-entering ctx restores the updated value of var.
+         with ctx:
+             assert var.get() == "updated"
+
+      .. versionadded:: 3.14
```
Comment (picnixz) [inline Doc/library/contextvars.rst]:
```suggestion
```

File: Doc/library/contextvars.rst [hunk 10]
```diff
@@ -144,21 +144,89 @@ Manual Context Management
...
+             assert var.get() == "updated"
+
+         # Exited ctx, so the observed value of var has reverted.
+         assert var.get() == "initial"
+         # But the updated value is still recorded in ctx.
+         assert ctx[var] == "updated"
+
+         # Re-entering ctx restores the updated value of var.
+         with ctx:
+             assert var.get() == "updated"
```
Comment (picnixz) [inline Doc/library/contextvars.rst]: I'd suggest using the same formulation as in `run()` where we don't use `assert` but rather comments.

File: Lib/test/test_context.py [hunk 11]
```diff
@@ -361,6 +362,74 @@ def sub(num):
             tp.shutdown()
         self.assertEqual(results, list(range(10)))
 
+    @isolated_context
+    def test_context_manager_works(self):
```
Comment (picnixz) [inline Lib/test/test_context.py]:
```suggestion
    def test_context_manager(self):
```

File: Lib/test/test_context.py [hunk 12]
```diff
@@ -361,6 +362,74 @@ def sub(num):
...
+    @threading_helper.requires_working_threading()
+    def test_context_manager_rejects_exit_from_different_thread(self):
+        ctx = contextvars.copy_context()
+        thread = threading.Thread(target=ctx.__enter__)
+        thread.start()
+        thread.join()
+        with self.assertRaises(RuntimeError):
+            ctx.__exit__(None, None, None)
+
+    def test_context_manager_rejects_recursive_enter_mgr_then_mgr(self):
```
Comment (picnixz) [inline Lib/test/test_context.py]:
```suggestion
    def test_context_manager_is_not_reentrant(self):
```

Re-entrancy in this case would mean that I can nest `with` statements, right? or is the term not adequate?

File: Lib/test/test_context.py [hunk 13]
```diff
@@ -361,6 +362,74 @@ def sub(num):
...
+        with contextvars.copy_context() as ctx:
+            with self.assertRaises(RuntimeError):
+                with ctx:
+                    pass
+
+    def test_context_manager_rejects_recursive_enter_mgr_then_run(self):
+        with contextvars.copy_context() as ctx:
+            with self.assertRaises(RuntimeError):
+                ctx.run(lambda: None)
```
Comment (picnixz) [inline Lib/test/test_context.py]: I think this test can be put in the above test, together with using `self.subTest` to indicate what kind of behaviour you're testing.

File: Lib/test/test_context.py [hunk 14]
```diff
@@ -361,6 +362,74 @@ def sub(num):
...
+
+    def test_context_manager_rejects_recursive_enter_run_then_mgr(self):
```
Comment (picnixz) [inline Lib/test/test_context.py]: Ditto.

File: Lib/test/test_context.py [hunk 15]
```diff
@@ -361,6 +362,74 @@ def sub(num):
...
+                    pass
+
+        ctx.run(fn)
+
+    def test_context_manager_rejects_noncurrent_exit(self):
+        with contextvars.copy_context() as ctx:
+            with contextvars.copy_context():
+                with self.assertRaises(RuntimeError):
+                    ctx.__exit__(None, None, None)
+
```
Comment (picnixz) [inline Lib/test/test_context.py]:
```suggestion
        with contextvars.copy_context() as outer_ctx:
            with contextvars.copy_context() as inner_ctx:
                self.assertIsNot(outer_ctx, inner_ctx)
                with self.assertRaises(RuntimeError):
                    outer_ctx.__exit__(None, None, None)
```

File: Misc/NEWS.d/next/Library/2022-11-21-01-24-46.gh-issue-99633.vhrNRe.rst [hunk 16]
```diff
@@ -0,0 +1 @@
+Add :term:`context manager` methods to :class:`contextvars.Context`.
```
Comment (picnixz) [inline Misc/NEWS.d/next/Library/2022-11-21-01-24-46.gh-issue-99633.vhrNRe.rst]:
```suggestion
Added support for the :term:`context management protocol` to
:class:`contextvars.Context` objects. Patch by Richard Hansen.
```

The feature is major enough (IMO) to have an additional What's New entry (not everyone reads the full changelog and this is a nice feature).

File: Python/context.c [hunk 17]
```diff
@@ -521,6 +520,69 @@ context_tp_contains(PyContext *self, PyObject *key)
...
 
 
+/*[clinic input]
+_contextvars.Context.__enter__
+
+Context manager enter.
+
+Automatically called by the 'with' statement.  Using the Context object as a
+context manager is an alternative to calling the Context.run() method.  Example
+usage:
```
Comment (picnixz) [inline Python/context.c]:
```suggestion
Automatically called by the 'with' statement.  Using the Context object as a
context manager is an alternative to calling the Context.run() method.

Example:
```

File: Python/context.c [hunk 18]
```diff
@@ -521,6 +520,69 @@ context_tp_contains(PyContext *self, PyObject *key)
...
+
+    var = contextvars.ContextVar('var')
+    var.set('initial')
+
+    with contextvars.copy_context():
+        # The current Context is a new copy of the previous Context.  Updating a
+        # context variable inside this 'with' statement only affects the new
+        # copy.
```
Comment (picnixz) [inline Python/context.c]:
```suggestion
        # Changes to context variables will be rolled back upon
        # exiting the `with` statement.
```

Maybe simplify the wording a bit?

File: Python/context.c [hunk 19]
```diff
@@ -521,6 +520,69 @@ context_tp_contains(PyContext *self, PyObject *key)
...
+        var.set('updated')
+        do_something_interesting()
+
+    # Now that the 'with' statement is done executing, the value of the 'var'
+    # context variable has reverted back to its value before the 'with'
+    # statement.
```
Comment (picnixz) [inline Python/context.c]:
```suggestion
    # The context variable value has been rolled back.
```

File: Python/context.c [hunk 20]
```diff
@@ -521,6 +520,69 @@ context_tp_contains(PyContext *self, PyObject *key)
...
+    assert var.get() == 'initial'
```
Comment (picnixz) [inline Python/context.c]:
```suggestion
    print(var.get())  # 'initial'
```

File: Python/context.c [hunk 21]
```diff
@@ -521,6 +520,69 @@ context_tp_contains(PyContext *self, PyObject *key)
...
+
+    var = contextvars.ContextVar('var')
+    var.set('initial')
+
+    with contextvars.copy_context():
+        # The current Context is a new copy of the previous Context.  Updating a
+        # context variable inside this 'with' statement only affects the new
+        # copy.
+        var.set('updated')
+        do_something_interesting()
```
Comment (picnixz) [inline Python/context.c]:
```suggestion
        print(var.get())  # 'updated'
```

File: Python/context.c (hunk 3 above)
Comment (picnixz) [reply Python/context.c:832]: Should we add a comment for people reading the code?

File: Python/context.c [hunk 22]
```diff
@@ -521,6 +520,69 @@ context_tp_contains(PyContext *self, PyObject *key)
...
+
+static PyObject *
+_contextvars_Context___exit___impl(PyContext *self, PyObject *exc_type,
+                                   PyObject *exc_val, PyObject *exc_tb)
+/*[clinic end generated code: output=4608fa9151f968f1 input=ff70cbbf6a112b1d]*/
+{
+    PyThreadState *ts = _PyThreadState_GET();
+    if (_PyContext_Exit(ts, (PyObject *)self)) {
+        return NULL;
+    }
```
Comment (picnixz) [inline Python/context.c]:
```suggestion
	if (PyContext_Exit((PyObject *)self) < 0) {
        return NULL;
    }
```

And if you want something even more concise, you can have something like:

```C
return PyContext_Exit((PyObject *)self) < 0 ? NULL : Py_None;
```

File: Python/context.c [hunk 23]
```diff
@@ -521,6 +520,69 @@ context_tp_contains(PyContext *self, PyObject *key)
...
+[clinic start generated code]*/
+
+static PyObject *
+_contextvars_Context___enter___impl(PyContext *self)
+/*[clinic end generated code: output=7374aea8983b777a input=22868e8274d3cd32]*/
+{
+    PyThreadState *ts = _PyThreadState_GET();
+    if (_PyContext_Enter(ts, (PyObject *)self)) {
+        return NULL;
+    }
```
Comment (picnixz) [inline Python/context.c]:
```suggestion
	if (PyContext_Enter((PyObject *)self) < 0) {
        return NULL;
    }
```

And the concise form:

```C
return PyContext_Enter((PyObject *)self) < 0 ? NULL : Py_NewRef(self);
```

**Comment by rhansen** [comment]
Thanks for the review, @picnixz!  I'll push a revision probably tomorrow.

File: Doc/library/contextvars.rst [hunk 24]
```diff
@@ -144,21 +144,89 @@ Manual Context Management
...
+   * Exiting from a Context that is not the current Context.
+
+   After exiting a Context, it can later be re-entered (from any thread).
+
+   Any changes to :class:`ContextVar` values via the :meth:`ContextVar.set`
+   method are recorded in the current Context.  The :meth:`ContextVar.get`
+   method returns the value associated with the current Context.  Thus, exiting
+   a Context effectively reverts any changes made to context variables while the
+   Context was entered.  (If desired, the values can be restored by re-entering
+   the Context.)
```
[the remark this answers is not in the document]
Comment (picnixz) [reply Doc/library/contextvars.rst:177]: I will split the suggestion into smaller ones. It's a bit hard to read if it's one huge chunk.

File: Doc/library/contextvars.rst [hunk 25]
```diff
@@ -144,21 +144,89 @@ Manual Context Management
...
    :func:`~contextvars.copy_context` function.
 
-   Every thread will have a different top-level :class:`~contextvars.Context`
-   object. This means that a :class:`ContextVar` object behaves in a similar
-   fashion to :func:`threading.local` when values are assigned in different
-   threads.
+   Each thread has its own effective stack of :class:`~contextvars.Context`
+   objects.  The *current Context* is the Context object at the top of the
+   current thread's stack.  All Context objects in the stacks are considered to
+   be *entered*.
```
Comment (picnixz) [inline Doc/library/contextvars.rst]:
```suggestion
   Each thread has its own effective stack of :class:`Context` objects.
   The *current context* is the :class:`!Context` object at the top of
   the current thread's stack.  All :class:`!Context` objects in the
   stacks are considered to be *entered*.
```

File: Doc/library/contextvars.rst [hunk 26]
```diff
@@ -144,21 +144,89 @@ Manual Context Management
...
+   current thread's stack.  All Context objects in the stacks are considered to
+   be *entered*.  
+   
+   *Entering* a Context, either by calling the
+   :meth:`Context.run` method or using the Context object as a :term:`context
+   manager`, pushes the Context onto the top of the current thread's stack,
+   making it the current Context.  *Exiting* from the current Context, either by
+   returning from the callback passed to :meth:`Context.run` or by exiting the
+   :keyword:`with` statement suite, pops the Context off of the top of the
+   stack, restoring the current Context to what it was before.
```
Comment (picnixz) [inline Doc/library/contextvars.rst]:
```suggestion
   *Entering* a context, either by calling the :meth:`~Context.run` method
   or using the context object as a :term:`context manager`, pushes the
   context onto the top of the current thread's stack, making it the
   current context.

   *Exiting* from the current context, either by returning from the callback
   passed to :meth:`~Context.run` or by exiting the :keyword:`with` statement
   suite, pops the contextoff of the top of the stack, restoring the current
   context to what it was before.
```

File: Doc/library/contextvars.rst [hunk 27]
```diff
@@ -144,21 +144,89 @@ Manual Context Management
...
+
+   Because each thread has its own Context stack, :class:`ContextVar` objects
```
Comment (picnixz) [inline Doc/library/contextvars.rst]:
```suggestion
   Since each thread has its own context stack, :class:`ContextVar` objects
```

File: Doc/library/contextvars.rst [hunk 28]
```diff
@@ -144,21 +144,89 @@ Manual Context Management
...
+   behave in a similar fashion to :func:`threading.local` when values are
+   assigned in different threads.
+
+   Attempting to do either of the following will raise a :exc:`RuntimeError`:
```
Comment (picnixz) [inline Doc/library/contextvars.rst]:
```suggestion
   Attempting to do either of the following raises a :exc:`RuntimeError`:
```

File: Doc/library/contextvars.rst [hunk 29]
```diff
@@ -144,21 +144,89 @@ Manual Context Management
...
+
+   * Entering an already entered Context.  (This includes Contexts entered in
+     other threads.)
+   * Exiting from a Context that is not the current Context.
```
Comment (picnixz) [inline Doc/library/contextvars.rst]:
```suggestion
   * Entering an already entered context, including contexts entered in
     other threads.
   * Exiting from a context that is not the current context.
```

File: Doc/library/contextvars.rst [hunk 30]
```diff
@@ -144,21 +144,89 @@ Manual Context Management
...
+
+   After exiting a Context, it can later be re-entered (from any thread).
```
Comment (picnixz) [inline Doc/library/contextvars.rst]:
```suggestion
   After exiting a context, it can later be re-entered (from any thread).
```

File: Doc/library/contextvars.rst (hunk 24 above)
Comment (picnixz) [inline Doc/library/contextvars.rst]:
```suggestion
   Any changes to :class:`ContextVar` values via the :meth:`ContextVar.set`
   method are recorded in the current context. The :meth:`ContextVar.get`
   method returns the value associated with the current context. Exiting
   a context effectively reverts any changes made to context variables while
   the context was entered (if needed, the values can be restored by re-entering
   the context).
```

File: Doc/library/contextvars.rst (hunk 10 above)
Comment (rhansen) [reply Doc/library/contextvars.rst]: Note the use of `.. testcode::`.  I want this example to be testable to prevent the implementation from drifting away from the documentation.  Would you prefer I use `print` and `.. testoutput::`?  I personally find `assert` easier to read.

File: Python/context.c (hunk 3 above)
Comment (rhansen) [reply Python/context.c:832]: Done.

File: Doc/library/contextvars.rst (hunk 7 above)
Comment (rhansen) [reply Doc/library/contextvars.rst]: "Context management protocol" is not defined in the glossary, which is why I linked to the PEP instead.

I could do ``:term:`context manager` protocol`` instead, although that terminology doesn't exactly match "context management protocol" used in the PEP.

Or I could add "context management protocol" to the glossary.

WDYT?

Comment (cjw296) [reply Doc/library/contextvars.rst]: Yep, let's do that.

File: Doc/library/contextvars.rst (hunk 10 above)
Comment (cjw296) [reply Doc/library/contextvars.rst]: Is that test code actually checked, though? I guess put differently: can you point me at the CPython CI that's actually executing this test code and checking it is as expected?

Comment (picnixz) [reply Doc/library/contextvars.rst]: That would be the doctest CI job. This is a sphinx directive for running doctest: https://www.sphinx-doc.org/en/master/usage/extensions/doctest.html#directive-testcode.

---

> Note the use of .. testcode::. I want this example to be testable to prevent the implementation from drifting away from the documentation. Would you prefer I use print and .. testoutput::? I personally find assert easier to read.

I think print() could be preferrable so that users, if they C/C the code, would also see the output directly. With a combination of `.. testoutput` that would be enough.

You can also change the rst for run() so that it also uses `testcode/testoutput`. Alternatively, you can use the `doctest` directive directly.

**Review by 1st1 — CHANGES_REQUESTED** [review]
I have a feeling this will not work at all with async/await. If that's the case then this API is a no-go. 

Please add a series of async/await tests with multiple context switches and contexts to show that this primitive is composable.

File: Doc/library/contextvars.rst [hunk 31]
```diff
@@ -144,21 +144,79 @@ Manual Context Management
...
+   context).
 
    Context implements the :class:`collections.abc.Mapping` interface.
 
+   .. versionadded:: 3.14
+      Added support for the :term:`!context management protocol` (:pep:`343`).
+      The value bound to the identifier given in the :keyword:`with` statement's
+      :keyword:`!as` clause (if present) is the :class:`!Context` object itself.
+
+      Example:
```
Comment (picnixz) [inline Doc/library/contextvars.rst]: The example should be out of the `.. versionchanged` section, ideally before (we should use `versionchanged` instead of `versionadded` since the latter is only when the object is being added, not when features are added).

File: Doc/library/contextvars.rst (hunk 10 above)
Comment (rhansen) [reply Doc/library/contextvars.rst]: I updated the `run` documentation to also use `testcode`/`testoutput`.

**Comment by rhansen** [comment]
> I have a feeling this will not work at all with async/await.

You are correct, @1st1.  I left a [comment in #99634 with more details](https://github.com/python/cpython/issues/99633#issuecomment-2350912468); let's continue the discussion there.  I'm going to mark this PR as draft until consensus is reached on a solution.
